components/org.wso2.carbon.identity.application.authenticator.biometric/src/main/java/org/wso2/carbon/identity/application/authenticator/biometric/BiometricAuthenticator.java

```diff
@@ -0,0 +1,161 @@
+                SIGNED_CHALLENGE));
+        return request.getParameter(BiometricAuthenticatorConstants.SIGNED_CHALLENGE) != null;
+    }
+
+    @Override
+    public String getContextIdentifier(javax.servlet.http.HttpServletRequest request) {
+
+        return request.getParameter(CONTEXT_KEY);
+    }
+
+    @Override
+    public String getName() {
+
+        return BiometricAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+                                                 AuthenticationContext context) throws AuthenticationFailedException {
+
+        AuthenticatedUser user = context.getSequenceConfig().getStepMap().
+                get(context.getCurrentStep() - 1).getAuthenticatedUser();
+        String username = user.getUserName();
+
+        Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+        String serverKey = authenticatorProperties.get(BiometricAuthenticatorConstants.SERVER_KEY);
+        String fcmUrl = authenticatorProperties.get(BiometricAuthenticatorConstants.FCM_URL);
+
+        String hostname = IdentityUtil.getHostName();
+        String serviceProviderName = context.getServiceProviderName();
+        String message = username + " is trying to log into " + serviceProviderName + " from " + hostname;
+        // TODO: 2019-11-20  support localization-future improvement.Include in DOCs.
+        String sessionDataKey = request.getParameter(CONTEXT_KEY);
+
+        UUID challenge = UUID.randomUUID();
+        String randomChallenge = challenge.toString();
+        context.setProperty(BIOMETRIC_AUTH_CHALLENGE, randomChallenge);
+
+        DeviceDAOImpl biometricDAO = DeviceDAOImpl.getInstance();
+        String deviceID = biometricDAO.getDeviceID(username).get(0);
+        log.info("challenge is: " + challenge);
+        log.info("SDK is : " + sessionDataKey);
+
+        FirebasePushNotificationSenderImpl pushNotificationSender = FirebasePushNotificationSenderImpl.getInstance();
+        pushNotificationSender.init(serverKey, fcmUrl);
+        pushNotificationSender.sendPushNotification(deviceID, message, randomChallenge, sessionDataKey);
+        try {
+            String pollingEndpoint = DOMAIN_NAME + WAIT_PAGE + "?sessionDataKey=";
+            String waitPage = pollingEndpoint + URLEncoder.encode(sessionDataKey, "UTF-8");
+            response.sendRedirect(waitPage);
+        } catch (IOException e) {
+            log.error("Error when trying to redirect to wait.jsp page", e);
+        }
+    }
+
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest httpServletRequest, HttpServletResponse
+            httpServletResponse, AuthenticationContext authenticationContext) throws AuthenticationFailedException {
+        log.info("now in process aauth method");
```

Comment:
Shall we remove this info log? If needed we can have a debug log with more readable content.